File: golang/rp_schedule_maint.go
///bin/true; exec /usr/bin/env go run "$0" "$@"

// Program: rp_schedule_maint.go
// Usage: go run rp_schedule_maint.go id 'start' 'end'
// Purpose: Go language sample client program for RackPing Monitoring API 2.0
// Version: 1.0
// Copyright: RackPing USA 2020
// Env: Go 1.13.14 or higher
// Returns: exit status is non-zero on failure
// Notes: first set envariables with: source ../set.sh

package main

import (
//      "bytes"
//      "encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
        "net/http/httputil"
	"os"
        "strconv"
        "strings"
	"time"
)

func debug(data []byte, err error) {
    if err == nil {
       fmt.Printf("%s\n\n", data)
    } else {
       log.Fatalf("%s\n\n", err)
    }
}

// global variables for communicating with callback redirect function redirectPolicyFunc()
var g_username string = ""
var g_password string = ""
var g_api_key  string = ""

func redirectPolicyFunc(req *http.Request, via []*http.Request) error {
   req.SetBasicAuth(g_username, g_password)
   req.Header.Set("app-key", g_api_key)
   return nil
}

func main() {
   var username string = os.Getenv("RP_USER")
   var password string = os.Getenv("RP_PASSWORD")
   var api_key  string = os.Getenv("RP_API_KEY")

   if username == "" {
      fmt.Printf("%s\n", "error: do 'source ../set.sh' first.")
      os.Exit(1)
   }

   var domain   string = os.Getenv("RP_DOMAIN")
   var base_url string = os.Getenv("RP_BASE_URL")
   var scheme   string = os.Getenv("RP_SCHEME")

   var Debug bool = (os.Getenv("RP_DEBUG") == "1")

   // see https://golang.org/pkg/time/ for how time casts work
   timeout, err := strconv.Atoi(os.Getenv("RP_TIMEOUT"))
   if err != nil {
      log.Fatal(err)
   }

   var url string = scheme + domain + base_url

   client := &http.Client{
      Timeout: time.Duration(timeout) * time.Second,
      CheckRedirect: redirectPolicyFunc,
   }

   g_username = username
   g_password = password
   g_api_key  = api_key

   {
      if (len(os.Args) < 4) {
         fmt.Fprintf(os.Stderr, "usage: %s id 'start' 'end'\n", os.Args[0])
         os.Exit(1)
      }

      id    := os.Args[1]
      start := strings.Replace(os.Args[2], " ", "%20", 1) 
      end   := strings.Replace(os.Args[3], " ", "%20", 1) 

      fmt.Printf("%s\n", "Enable maintenance window for one check")

      req, err := http.NewRequest("PUT", url + "/checks/" + id + "?start_maintenance=" + start + "&end_maintenance=" + end, nil)
      req.SetBasicAuth(g_username, g_password)
      req.Header.Set("app-key", api_key)

      if Debug { debug(httputil.DumpRequestOut(req, true)) }

      res, err := client.Do(req)

      if err != nil {
         log.Fatal(err)
         debug(httputil.DumpResponse(res, true))
      } else {
         defer res.Body.Close()
         if Debug { debug(httputil.DumpResponse(res, true)) }
         body, err := ioutil.ReadAll(res.Body)
         if err == nil {
            fmt.Printf("%s\n", body)
         }
      }
   }

   os.Exit(0)
}

File: perl/rp_schedule_maint.pl
#!/usr/bin/perl

# Program: rp_schedule_maint.pl
# Usage: ./rp_schedule_maint.pl id "start" "end"
# Purpose: perl language sample client program for RackPing Monitoring API 2.0
# Version: 1.0
# Copyright: RackPing USA 2020
# Env: Perl5
# Returns: exit status is non-zero on failure
# Note: First set the envariables with: source ../set.sh

use strict;
use diagnostics;

use MIME::Base64 qw(encode_base64);
use LWP::UserAgent;

   use constant TIMEOUT_SEC => 5.0;

   my $url = $ENV{'RP_SCHEME'} . $ENV{'RP_DOMAIN'} . $ENV{'RP_BASE_URL'};

   my $api_key  = $ENV{'RP_API_KEY'}  // '';
   my $user     = $ENV{'RP_USER'}     // '';
   my $password = $ENV{'RP_PASSWORD'} // '';
   my $timeout  = $ENV{'RP_TIMEOUT'}  || TIMEOUT_SEC;
   my $useragent = $ENV{'RP_USERAGENT'} // '';
   my $DEBUG    = $ENV{'RP_DEBUG'}    || 0;

   if ($api_key eq '') {
      print "error: do source ../set.sh to set the envariables.\n";
      exit 1;
   }

   if (scalar @ARGV < 3) {
      print qq[usage: $0 id "start" "end"\n];
      exit 1;
   }

### start of user settings

   my $check  = shift // '';
   my $start  = shift // '';
   my $end    = shift // '';

   $start =~ s/ /%20/g;
   $end   =~ s/ /%20/g;

### end of user settings

# Create a user agent object
   my $ua = LWP::UserAgent->new;

   $ua->from('rackping@example.com');
   $ua->timeout($timeout); # 411 - not reliable for https requests?

   my @headers = (
      'Accept'         => 'application/json',
      'Accept-Charset' => 'utf-8',
      'App-key'        => "$api_key",
      'Authorization'  => 'Basic ' . encode_base64("$user:$password", ''),
      'Content-type'   => 'application/json',
      'User-Agent'     => $useragent,
   );

   $ua->default_header(@headers);

# Create a request for one check. The format is MySQL, or ISO without the 'T' and no fractional seconds.

   my $req = HTTP::Request->new(PUT => $url . "/checks/$check?start_maintenance=$start&end_maintenance=$end");

# Pass request to the user agent and get a response back
   my $res = $ua->request($req);

   print STDERR "info: start maintenance on one check\n" if $DEBUG;

# Check the outcome of the response
   if ($res->is_success) {
      print $res->content, "\n";
   }
   else {
      print $res->status_line, "\n";
      exit 1;
   }

   exit 0;

